testsuite: refactor mtest to share code between c and cxx tests

hzhou opened the pull request:
Instead of duplicate code between `mtest.c` and `mtest.cxx`, use `mtest.inl` include file to share code.

`.inl` extension has been used in many other projects, referring to `inline`, but has become a convention for source/object level include files.

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
@pavanbalaji This PR is ready for your review.

pavanbalaji commented:
What's the reason for the `.inl` file?  Can't we keep that as a `.c` file and link it to both the C and C++ tests?

hzhou commented:
> What's the reason for the `.inl` file? Can't we keep that as a `.c` file and link it to both the C and C++ tests?

It is a simpler. It almost works the same as before just without duplicating the code. What is the reason to make it into separate compile unit?

pavanbalaji commented:
> > What's the reason for the `.inl` file? Can't we keep that as a `.c` file and link it to both the C and C++ tests?
> 
> It is a simpler. It almost works the same as before just without duplicating the code. What is the reason to make it into separate compile unit?

My reasoning was to avoid including C code, which I feel is more of a role for `.h` files.

hzhou commented:
It is C++ code as far as compiler concerns :)

pavanbalaji commented:
> It is C++ code as far as compiler concerns :)

I understand how the compiler sees it.  But that was not what I was saying.

pavanbalaji commented:
If the concern is only about writing the build system, I can help with it.  It's not as difficult as you might think.  In fact, some would even consider that simpler, instead of the `.inl` includes.  :-)

hzhou commented:
We want the code be treated by compiler in its proper C++ way, right? Otherwise, we have to worry about the difference between C and C++ compilers, even though that is being kept minimal.

Also by commenting on which is a simpler solution does not mean the same as difficulty.

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented on test/mpi/include/mpitestcxx.h:
```diff
@@ -8,28 +8,27 @@
-/*
- * Init and finalize test
- */
 void MTest_Init(void);
```
`gnuindent` is having trouble dealing here, keep concatenating `void MTest_Init(void);` to the previous comment line.

hzhou commented:
test:mpich/quick

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
@pavanbalaji The tests seems to be good. It is ready for your review.

pavanbalaji commented on test/mpi/Makefile.mtest:
```diff
@@ -16,20 +16,18 @@
 # AM_CPPFLAGS are used for C++ code as well
 AM_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
-LDADD = $(top_builddir)/util/mtest.o
+LDADD = $(top_builddir)/util/mtest.$(OBJEXT) $(top_builddir)/util/mtest_common.$(OBJEXT)
```
Changing `mtest.o` to `mtest.$(OBJEXT)` is a bug-fix and should be separated into a different commit.

pavanbalaji commented on test/mpi/Makefile_cxx.mtest:
```diff
@@ -12,19 +12,17 @@
 
 AM_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
-LDADD = $(top_builddir)/cxx/util/mtest.o
+LDADD = $(top_builddir)/util/mtest_cxx.$(OBJEXT) $(top_builddir)/util/mtest_common.$(OBJEXT)
```
Changing mtest_cxx.o to mtest_cxx.$(OBJEXT) is a bug-fix and should be separated into a different commit.

pavanbalaji commented on test/mpi/include/mpitest.h:
```diff
@@ -7,10 +7,10 @@
 #ifndef MPITEST_H_INCLUDED
 #define MPITEST_H_INCLUDED
 
-#include <string.h>
```
Removal of unnecessary headers should be a different commit.

pavanbalaji commented on test/mpi/include/mpitestcxx.h:
```diff
@@ -8,28 +8,27 @@
+#include <iostream.h>
 #endif
-
-#include <string.h>
```
Removal of unnecessary headers should be a different commit.

pavanbalaji commented on test/mpi/include/mpitestcxx.h:
```diff
@@ -8,28 +8,27 @@
-#include <string.h>
-/*
- * Init and finalize test
- */
```
It's OK to remove this comment, but it should be separated into a different commit.

pavanbalaji commented on test/mpi/util/mtest.c:
```diff
@@ -30,7 +30,6 @@
 #endif
 #include <errno.h>
 
-
```
Unrelated line removal.

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
@pavanbalaji I have split the commits out. I also added a patch for whitespace-checker as it does not deal with c++ header. It is ready for your another review.
